Abstract Dialog Test class

I need the capability to convert a .dialog file into an OWL file using the JenaBasedDialogModelProcessor from a JUnit test. For SADL, this is accomplished through calls to assertValidatesTo in AbstractSadlTest. I don't know the easiest/best way to so this for DIALOG files. I suspect that you, @kittaakos , could pretty quickly provide this capability? In a fraction of the time it would take me to figure out how to accomplish it. This is need to provide end-to-end JUnit tests/standalone demos.

I cannot build it. SADL, as of today from the `development` branch, has `3.2.1-SNAPSHOT` version. DARPA requires `3.2.0-SNAPSHOT` from the target platform. Which SADL branch should I use? Please note, I had to wipe my local Maven cache, I do not have SADL `3.2.0` available there.

Local Maven:
```
Akoss-MacBook-Pro:com.ge.research.sadl akos.kitta$ pwd
/Users/akos.kitta/.m2/repository/com/ge/research/sadl/com.ge.research.sadl
Akoss-MacBook-Pro:com.ge.research.sadl akos.kitta$ ls -al
total 8
drwxr-xr-x   4 akos.kitta  staff   128 Sep 10 08:31 .
drwxr-xr-x  32 akos.kitta  staff  1024 Sep 10 08:33 ..
drwxr-xr-x   8 akos.kitta  staff   256 Sep 10 08:31 3.2.1-SNAPSHOT
-rw-r--r--   1 akos.kitta  staff   300 Sep 10 08:31 maven-metadata-local.xml
Akoss-MacBook-Pro:com.ge.research.sadl akos.kitta$ 
```

Error:
```
[INFO] Scanning for projects...
[ERROR] Internal error: java.lang.RuntimeException: Could not resolve target platform specification artifact com.ge.research.sadl:com.ge.research.sadl.target:target:3.2.0-SNAPSHOT -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: Could not resolve target platform specification artifact com.ge.research.sadl:com.ge.research.sadl.target:target:3.2.0-SNAPSHOT
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:122)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:955)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: java.lang.RuntimeException: Could not resolve target platform specification artifact com.ge.research.sadl:com.ge.research.sadl.target:target:3.2.0-SNAPSHOT
    at org.eclipse.tycho.core.resolver.DefaultTargetPlatformConfigurationReader.addTargetArtifact (DefaultTargetPlatformConfigurationReader.java:389)
    at org.eclipse.tycho.core.resolver.DefaultTargetPlatformConfigurationReader.setTarget (DefaultTargetPlatformConfigurationReader.java:342)
    at org.eclipse.tycho.core.resolver.DefaultTargetPlatformConfigurationReader.getTargetPlatformConfiguration (DefaultTargetPlatformConfigurationReader.java:75)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.setupProject (DefaultTychoResolver.java:87)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.afterProjectsRead (TychoMavenLifecycleParticipant.java:90)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:267)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:955)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/InternalErrorException
```

I still cannot compile. How do you set up DARPA with SADL source? Thanks!

```
java.lang.RuntimeException: Problems running workflow com.ge.research.sadl.darpa.aske.GenerateDialog: org.eclipse.emf.ecore.xmi.IllegalValueException: Value '2.16' is not legal. (platform:/resource/com.ge.research.sadl/model/generated/SADL.genmodel, 6, 73)
	at org.eclipse.emf.mwe2.launch.runtime.Mwe2Runner.run(Mwe2Runner.java:104)
	at org.eclipse.emf.mwe2.launch.runtime.Mwe2Runner.run(Mwe2Runner.java:62)
	at org.eclipse.emf.mwe2.launch.runtime.Mwe2Runner.run(Mwe2Runner.java:52)
	at org.eclipse.emf.mwe2.launch.runtime.Mwe2Launcher.run(Mwe2Launcher.java:78)
	at org.eclipse.emf.mwe2.launch.runtime.Mwe2Launcher.main(Mwe2Launcher.java:36)
Caused by: org.eclipse.emf.ecore.resource.impl.ResourceSetImpl$1DiagnosticWrappedException: org.eclipse.emf.ecore.xmi.IllegalValueException: Value '2.16' is not legal. (platform:/resource/com.ge.research.sadl/model/generated/SADL.genmodel, 6, 73)
	at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.handleDemandLoadException(ResourceSetImpl.java:319)
	at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.demandLoadHelper(ResourceSetImpl.java:278)
	at org.eclipse.xtext.resource.XtextResourceSet.getResource(XtextResourceSet.java:265)
	at org.eclipse.xtext.xtext.generator.XtextGeneratorResourceSetInitializer.loadResource(XtextGeneratorResourceSetInitializer.java:51)
	at org.eclipse.xtext.xtext.generator.XtextGeneratorResourceSetInitializer.lambda$initialize$0(XtextGeneratorResourceSetInitializer.java:41)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.eclipse.xtext.xtext.generator.XtextGeneratorResourceSetInitializer.initialize(XtextGeneratorResourceSetInitializer.java:43)
	at org.eclipse.xtext.xtext.generator.XtextGeneratorLanguage.initialize(XtextGeneratorLanguage.java:218)
	at org.eclipse.xtext.xtext.generator.StandardLanguage.initialize(StandardLanguage.java:189)
	at org.eclipse.xtext.xtext.generator.XtextGenerator.initialize(XtextGenerator.java:185)
	at org.eclipse.xtext.xtext.generator.XtextGenerator.checkConfigurationInternal(XtextGenerator.java:139)
	at org.eclipse.emf.mwe.core.lib.AbstractWorkflowComponent2.checkConfiguration(AbstractWorkflowComponent2.java:21)
	at org.eclipse.emf.mwe.core.lib.Mwe2Bridge.preInvoke(Mwe2Bridge.java:65)
	at org.eclipse.emf.mwe.core.lib.AbstractWorkflowComponent.preInvoke(AbstractWorkflowComponent.java:197)
	at org.eclipse.emf.mwe2.runtime.workflow.AbstractCompositeWorkflowComponent.preInvoke(AbstractCompositeWorkflowComponent.java:29)
	at org.eclipse.emf.mwe2.runtime.workflow.Workflow.run(Workflow.java:18)
	at org.eclipse.emf.mwe2.launch.runtime.Mwe2Runner.run(Mwe2Runner.java:102)
	... 4 more
Caused by: org.eclipse.emf.ecore.xmi.IllegalValueException: Value '2.16' is not legal. (platform:/resource/com.ge.research.sadl/model/generated/SADL.genmodel, 6, 73)
	at org.eclipse.emf.ecore.xmi.impl.XMLHandler.setFeatureValue(XMLHandler.java:2715)
	at org.eclipse.emf.ecore.xmi.impl.XMLHandler.setAttribValue(XMLHandler.java:2769)
	at org.eclipse.emf.ecore.xmi.impl.SAXXMIHandler.handleObjectAttribs(SAXXMIHandler.java:79)
	at org.eclipse.emf.ecore.xmi.impl.XMLHandler.createObjectFromFactory(XMLHandler.java:2247)
	at org.eclipse.emf.ecore.xmi.impl.XMLHandler.createObjectByType(XMLHandler.java:1384)
	at org.eclipse.emf.ecore.xmi.impl.XMLHandler.createTopObject(XMLHandler.java:1522)
	at org.eclipse.emf.ecore.xmi.impl.XMLHandler.processElement(XMLHandler.java:1044)
	at org.eclipse.emf.ecore.xmi.impl.XMIHandler.processElement(XMIHandler.java:78)
	at org.eclipse.emf.ecore.xmi.impl.XMLHandler.startElement(XMLHandler.java:1026)
	at org.eclipse.emf.ecore.xmi.impl.XMLHandler.startElement(XMLHandler.java:720)
	at org.eclipse.emf.ecore.xmi.impl.XMIHandler.startElement(XMIHandler.java:190)
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.startElement(AbstractSAXParser.java:509)
	at com.sun.org.apache.xerces.internal.impl.dtd.XMLDTDValidator.startElement(XMLDTDValidator.java:745)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanStartElement(XMLDocumentFragmentScannerImpl.java:1359)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$ContentDriver.scanRootElementHook(XMLDocumentScannerImpl.java:1289)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:3132)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:852)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:602)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:505)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:842)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:771)
	at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1213)
	at com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:643)
	at com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl.parse(SAXParserImpl.java:327)
	at org.eclipse.emf.ecore.xmi.impl.XMLLoadImpl.load(XMLLoadImpl.java:175)
	at org.eclipse.emf.ecore.xmi.impl.XMLResourceImpl.doLoad(XMLResourceImpl.java:261)
	at org.eclipse.emf.ecore.resource.impl.ResourceImpl.load(ResourceImpl.java:1563)
	at org.eclipse.emf.ecore.resource.impl.ResourceImpl.load(ResourceImpl.java:1342)
	at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.demandLoad(ResourceSetImpl.java:259)
	at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.demandLoadHelper(ResourceSetImpl.java:274)
	... 19 more
Caused by: java.lang.IllegalArgumentException: The value '2.16' is not a valid enumerator of 'GenRuntimeVersion'
	at org.eclipse.emf.codegen.ecore.genmodel.impl.GenModelFactoryImpl.createGenRuntimeVersionFromString(GenModelFactoryImpl.java:406)
	at org.eclipse.emf.codegen.ecore.genmodel.impl.GenModelFactoryImpl.createFromString(GenModelFactoryImpl.java:114)
	at org.eclipse.emf.ecore.xmi.impl.XMLHelperImpl.createFromString(XMLHelperImpl.java:1615)
	at org.eclipse.emf.ecore.xmi.impl.XMLHelperImpl.setValue(XMLHelperImpl.java:1156)
	at org.eclipse.emf.ecore.xmi.impl.XMLHandler.setFeatureValue(XMLHandler.java:2710)
	... 49 more
```

--- a/com.ge.research.sadl.darpa.aske.dialog.parent/com.ge.research.sadl.darpa.aske.dialog.tests/META-INF/MANIFEST.MF
+++ b/com.ge.research.sadl.darpa.aske.dialog.parent/com.ge.research.sadl.darpa.aske.dialog.tests/META-INF/MANIFEST.MF
@@ -14,7 +14,9 @@ Require-Bundle: com.ge.research.sadl.darpa.aske.dialog,
  com.ge.research.sadl;bundle-version="3.2.0",
  com.google.gson,
  org.eclipse.jdt.core,
- com.ge.research.sadl.jena
+ com.ge.research.sadl.jena,
+ com.ge.research.sadl.tests,
+ com.ge.research.jena;bundle-version="3.2.0"
 Bundle-RequiredExecutionEnvironment: JavaSE-1.8
 Export-Package: com.ge.research.sadl.darpa.aske.tests;x-internal=true
 Import-Package: com.ge.research.sadl.reasoner,

--- a/com.ge.research.sadl.darpa.aske.dialog.parent/com.ge.research.sadl.darpa.aske.dialog.tests/src/com/ge/research/sadl/darpa/aske/tests/AbstractDialogTest.xtend
+++ b/com.ge.research.sadl.darpa.aske.dialog.parent/com.ge.research.sadl.darpa.aske.dialog.tests/src/com/ge/research/sadl/darpa/aske/tests/AbstractDialogTest.xtend
@@ -1,0 +1,39 @@
+package com.ge.research.sadl.darpa.aske.tests
+
+import com.ge.research.sadl.jena.IJenaBasedModelProcessor
+import com.ge.research.sadl.model.gp.Rule
+import com.ge.research.sadl.model.gp.SadlCommand
+import com.ge.research.sadl.tests.AbstractSadlTest
+import com.ge.research.sadl.tests.SadlTestAssertions
+import com.hp.hpl.jena.ontology.OntModel
+import java.util.List
+import org.eclipse.emf.ecore.resource.Resource
+import org.eclipse.xtext.resource.XtextResource
+import org.eclipse.xtext.testing.InjectWith
+import org.eclipse.xtext.testing.XtextRunner
+import org.eclipse.xtext.validation.Issue
+import org.junit.runner.RunWith
+
+@RunWith(XtextRunner)
+@InjectWith(DialogInjectorProvider)
+abstract class AbstractDialogTest extends AbstractSadlTest {
+
+	protected def XtextResource dialog(CharSequence seq) {
+		return resource(seq, 'dialog') as XtextResource;
+	}
+
+	protected def Resource assertValidatesTo(CharSequence code,
+		(OntModel, List<Issue>, IJenaBasedModelProcessor)=>void assertions) {
+
+		return SadlTestAssertions.assertValidatesTo(code.dialog, [ ontModel, rules, commands, issues, processor |
+			assertions.apply(ontModel, issues, processor)
+		]);
+	}
+
+	override Resource assertValidatesTo(CharSequence code,
+		(OntModel, List<Rule>, List<SadlCommand>, List<Issue>, IJenaBasedModelProcessor)=>void assertions) {
+
+		return SadlTestAssertions.assertValidatesTo(code.dialog, assertions);
+	}
+
+}

--- a/com.ge.research.sadl.darpa.aske.dialog.parent/com.ge.research.sadl.darpa.aske.dialog.tests/src/com/ge/research/sadl/darpa/aske/tests/DialogTest.xtend
+++ b/com.ge.research.sadl.darpa.aske.dialog.parent/com.ge.research.sadl.darpa.aske.dialog.tests/src/com/ge/research/sadl/darpa/aske/tests/DialogTest.xtend
@@ -1,0 +1,237 @@
+package com.ge.research.sadl.darpa.aske.tests
+
+import org.eclipse.xtext.diagnostics.Severity
+import org.junit.Test
+
+import static org.junit.Assert.assertNotNull
+import static org.junit.Assert.assertTrue
+
+class DialogTest extends AbstractDialogTest {
+	
+	@Test
+	def void dummy_test() {
+		'''
+			uri  "http://sadl.org/Test.sadl" alias Test.
+			
+			Rock is
+			
+			Person is a class described by age with values of type int.
+			ChildrenList is a type of Person List length 1-*.
+			children describes Person with values of type ChildrenList.
+			children of Person has at most 1 value.
+			
+			PersonList is a type of Person List.
+			PersonListList is a type of PersonList List.
+			
+			foo describes Person with values of type PersonListList.
+			bar describes Person with values of type Person List length 1-4.
+			bar of Person only has values of type Person List.
+			bar of Person only has values of type Person List length 1-4.
+			bar of Person has at least one value of type Person List length 1-4.
+			bar of Person has at least 1 value of type Person List length 1-4.
+			bar of Person has at most 2 value of type Person List length 1-4.
+			
+			
+			Rule R1: if x is a Person and
+					x has bar y and 
+					y is a Person List //length 1-4
+					z is a Person List length 1-4
+			then print("Hurray!"). //x has age 50.
+		'''.sadl
+		
+		'''
+		uri "http://baseUri".
+		import "http://sadl.org/SpeedOfSound.sadl".
+		
+		
+		CM: "Parsing code file 'C:/Users/200005201/sadl3-master6/git/DARPA-ASKE-TA1/Ontology/M5/ExtractedModels/Sources/Mach.java'".
+		CM: "The following methods doing computation were found in the extraction:".
+		CM: External Mach.CAL_TT(double T, double M, double G, double Q) returns double: "http://com.ge.research.sadl.darpa.aske.answer/Mach_java#Mach.CAL_TT".
+		
+		CM: Mach.CAL_TT has expression (a Script with language Java, with script 
+		"public double CAL_TT(double T, double M, double G, double Q) {
+		    double TT = T / TQTT(M, G);
+		    double EPS = 0.00001;
+		    double Z = 1;
+		    double Z2 = 0;
+		    double TT2 = 0;
+		    double TT1;
+		    double Z1;
+		    double i = 1;
+		    while (Math.abs(Z) > EPS) {
+		        Z = Math.pow(M, 2) - 2 * TT / CAL_GAM(T, G, Q) / T * (G / (G - 1) * (1 - T / TT) + Q / TT * (1 / (Math.exp(Q / TT) - 1) - 1 / (Math.exp(Q / T) - 1)));
+		        if (i == 1) {
+		            Z2 = Z;
+		            TT2 = TT;
+		            TT = TT2 + 100;
+		            i = 2;
+		        } else {
+		            TT1 = TT2;
+		            Z1 = Z2;
+		            TT2 = TT;
+		            Z2 = Z;
+		            TT = TT2 - Z2 * (TT2 - TT1) / (Z2 - Z1);
+		        }
+		    }
+		    if (M <= .1)
+		        TT = T / TQTT(M, G);
+		    return TT;
+		}"
+		), has expression (a Script with language Python, with script 
+		"#!/usr/bin/env python
+		\"\"\" generated source for module inputfile \"\"\"
+		class Mach(object):
+		    \"\"\" generated source for class Mach \"\"\"
+		    def CAL_TT(self, T, M, G, Q):
+		        \"\"\" generated source for method CAL_TT \"\"\"
+		        TT = T / TQTT(M, G)
+		        EPS = 0.00001
+		        Z = 1
+		        Z2 = 0
+		        TT2 = 0
+		        TT1 = float()
+		        Z1 = float()
+		        i = 1
+		        while Math.abs(Z) > EPS:
+		            Z = Math.pow(M, 2) - 2 * TT / CAL_GAM(T, G, Q) / T * (G / (G - 1) * (1 - T / TT) + Q / TT * (1 / (Math.exp(Q / TT) - 1) - 1 / (Math.exp(Q / T) - 1)))
+		            if i == 1:
+		                Z2 = Z
+		                TT2 = TT
+		                TT = TT2 + 100
+		                i = 2
+		            else:
+		                TT1 = TT2
+		                Z1 = Z2
+		                TT2 = TT
+		                Z2 = Z
+		                TT = TT2 - Z2 * (TT2 - TT1) / (Z2 - Z1)
+		        if M <= 0.1:
+		            TT = T / TQTT(M, G)
+		        return TT
+		
+		").
+		CM: External Mach.CAL_SOS(double T, double G, double R, double Q) returns double: "http://com.ge.research.sadl.darpa.aske.answer/Mach_java#Mach.CAL_SOS".
+		
+		CM: Mach.CAL_SOS has expression (a Script with language Java, with script 
+		"public double CAL_SOS(double T, double G, double R, double Q) {
+		    double WOW = 1 + (G - 1) / (1 + (G - 1) * Math.pow((Q / T), 2) * Math.exp(Q / T) / Math.pow((Math.exp(Q / T) - 1), 2));
+		    return (Math.sqrt(32.174 * T * R * WOW));
+		}"
+		), has expression (a Script with language Python, with script 
+		"#!/usr/bin/env python
+		\"\"\" generated source for module inputfile \"\"\"
+		class Mach(object):
+		    \"\"\" generated source for class Mach \"\"\"
+		    def CAL_SOS(self, T, G, R, Q):
+		        \"\"\" generated source for method CAL_SOS \"\"\"
+		        WOW = 1 + (G - 1) / (1 + (G - 1) * Math.pow((Q / T), 2) * Math.exp(Q / T) / Math.pow((Math.exp(Q / T) - 1), 2))
+		        return (Math.sqrt(32.174 * T * R * WOW))
+		
+		").
+		CM: External Mach.CAL_T(double TT, double M, double G, double Q) returns double: "http://com.ge.research.sadl.darpa.aske.answer/Mach_java#Mach.CAL_T".
+		
+		CM: Mach.CAL_T has expression (a Script with language Java, with script 
+		"double CAL_T(double TT, double M, double G, double Q) {
+		    double T = TT * TQTT(M, G);
+		    double EPS = 0.00001;
+		    double Z = 1;
+		    double Z2 = 0;
+		    double T2 = 0;
+		    double T1;
+		    double Z1;
+		    double i = 1;
+		    while (Math.abs(Z) > EPS) {
+		        Z = Math.pow(M, 2) - 2 * TT / CAL_GAM(T, G, Q) / T * (G / (G - 1) * (1 - T / TT) + Q / TT * (1 / (Math.exp(Q / TT) - 1) - 1 / (Math.exp(Q / T) - 1)));
+		        if (i == 1) {
+		            Z2 = Z;
+		            T2 = T;
+		            T = T2 + 100;
+		            i = 2;
+		        } else {
+		            T1 = T2;
+		            Z1 = Z2;
+		            T2 = T;
+		            Z2 = Z;
+		            T = T2 - Z2 * (T2 - T1) / (Z2 - Z1);
+		        }
+		    }
+		    if (M <= .1)
+		        T = TT * TQTT(M, G);
+		    return T;
+		}"
+		), has expression (a Script with language Python, with script 
+		"#!/usr/bin/env python
+		\"\"\" generated source for module inputfile \"\"\"
+		class Mach(object):
+		    \"\"\" generated source for class Mach \"\"\"
+		    def CAL_T(self, TT, M, G, Q):
+		        \"\"\" generated source for method CAL_T \"\"\"
+		        T = TT * TQTT(M, G)
+		        EPS = 0.00001
+		        Z = 1
+		        Z2 = 0
+		        T2 = 0
+		        T1 = float()
+		        Z1 = float()
+		        i = 1
+		        while Math.abs(Z) > EPS:
+		            Z = Math.pow(M, 2) - 2 * TT / CAL_GAM(T, G, Q) / T * (G / (G - 1) * (1 - T / TT) + Q / TT * (1 / (Math.exp(Q / TT) - 1) - 1 / (Math.exp(Q / T) - 1)))
+		            if i == 1:
+		                Z2 = Z
+		                T2 = T
+		                T = T2 + 100
+		                i = 2
+		            else:
+		                T1 = T2
+		                Z1 = Z2
+		                T2 = T
+		                Z2 = Z
+		                T = T2 - Z2 * (T2 - T1) / (Z2 - Z1)
+		        if M <= 0.1:
+		            T = TT * TQTT(M, G)
+		        return T
+		
+		").
+		CM: External Mach.CAL_GAM(double T, double G, double Q) returns double: "http://com.ge.research.sadl.darpa.aske.answer/Mach_java#Mach.CAL_GAM".
+		
+		CM: Mach.CAL_GAM has expression (a Script with language Java, with script 
+		"public double CAL_GAM(double T, double G, double Q) {
+		    return (1 + (G - 1) / (1 + (G - 1) * (Math.pow((Q / T), 2) * Math.exp(Q / T) / Math.pow((Math.exp(Q / T) - 1), 2))));
+		}"
+		), has expression (a Script with language Python, with script 
+		"#!/usr/bin/env python
+		\"\"\" generated source for module inputfile \"\"\"
+		class Mach(object):
+		    \"\"\" generated source for class Mach \"\"\"
+		    def CAL_GAM(self, T, G, Q):
+		        \"\"\" generated source for method CAL_GAM \"\"\"
+		        return (1 + (G - 1) / (1 + (G - 1) * (Math.pow((Q / T), 2) * Math.exp(Q / T) / Math.pow((Math.exp(Q / T) - 1), 2))))
+		
+		").
+		CM: External Mach.TQTT(double M, double G) returns double: "http://com.ge.research.sadl.darpa.aske.answer/Mach_java#Mach.TQTT".
+		
+		CM: Mach.TQTT has expression (a Script with language Java, with script 
+		"public double TQTT(double M, double G) {
+		    return Math.pow((1 + (G - 1) / 2 * Math.pow(M, 2)), -1);
+		}"
+		), has expression (a Script with language Python, with script 
+		"#!/usr/bin/env python
+		\"\"\" generated source for module inputfile \"\"\"
+		class Mach(object):
+		    \"\"\" generated source for class Mach \"\"\"
+		    def TQTT(self, M, G):
+		        \"\"\" generated source for method TQTT \"\"\"
+		        return Math.pow((1 + (G - 1) / 2 * Math.pow(M, 2)), -1)
+		
+		").
+		CM: "Would you like to save the extracted model (Mach.java.owl) in SADL format"?
+		
+		Build CAL_SOS.
+		CM: "Failed: Unable to find model 'http://sadl.org/SpeedOfSound.sadl#CAL_SOS' in code model".
+		'''.assertValidatesTo [ ontModel, issues, processor |
+			assertNotNull(ontModel)
+			assertTrue(issues.filter[severity === Severity.ERROR].empty)
+		]
+	}
+	
+}

